// @progress/kendo-angular-buttons/dist/es2015/buttongroup/buttongroup.component.js
import * as tslib_1 from "tslib";
import { ButtonDirective } from '../button/button.directive';
import { Component, Input, ContentChildren, QueryList, HostBinding, HostListener, isDevMode, Optional, Inject } from '@angular/core';
import { RTL } from '@progress/kendo-angular-l10n';
import { KendoButtonService } from '../button/button.service';
import { isPresent } from '../util';
import { Keys } from '../navigation/keys';
/**
 * @hidden
 */
const ariaChecked = 'aria-checked';
/**
 * @hidden
 */
const role = 'role';
/**
 * @hidden
 */
const tabindex = 'tabindex';
/**
 * Represents the Kendo UI ButtonGroup component for Angular.
 */
let ButtonGroupComponent = class ButtonGroupComponent {
    constructor(service, rtl) {
        this.service = service;
        /**
         * By default, the selection mode of the ButtonGroup is set to `multiple`.
         */
        this.selection = 'multiple';
        /**
         * Changes the visual appearance by using alternative styling options.
         * The available values are:
         * * `bare`
         * * `flat`
         * * `outline`
         *
         * The `look` property of the ButtonGroup takes precedence over the `look` property
         * of the individual buttons that are part of the group.
         */
        this.look = 'default';
        this.direction = rtl ? 'rtl' : 'ltr';
    }
    get wrapperClass() {
        return true;
    }
    get disabledClass() {
        return this.disabled;
    }
    get stretchedClass() {
        return !!this.width;
    }
    get isBare() {
        return this.look === 'bare';
    }
    get isFlat() {
        return this.look === 'flat';
    }
    get isOutline() {
        return this.look === 'outline';
    }
    get getRole() {
        return this.isSelectionSingle() ? 'radiogroup' : 'group';
    }
    get dir() {
        return this.direction;
    }
    get ariaDisabled() {
        return this.disabled;
    }
    get wrapperWidth() {
        return this.width;
    }
    /**
     * @hidden
     */
    keydown(event) {
        if (this.isSelectionSingle()) {
            const selectedIndex = this.buttons.toArray().findIndex(current => current.selected);
            const firstIndex = 0;
            const lastIndex = this.buttons.length - 1;
            if (selectedIndex !== undefined) {
                if (event.keyCode === Keys.right && selectedIndex < lastIndex) {
                    this.deactivate(this.buttons.filter(current => current.selected));
                    this.activate(this.buttons.filter((_current, index) => {
                        return index === selectedIndex + 1;
                    }));
                }
                if (event.keyCode === Keys.left && selectedIndex > firstIndex) {
                    this.deactivate(this.buttons.filter(current => current.selected));
                    this.activate(this.buttons.filter((_current, index) => {
                        return index === selectedIndex - 1;
                    }));
                }
            }
        }
    }
    ngOnInit() {
        this.subscription = this.service.buttonClicked$.subscribe((button) => {
            if (this.isSelectionSingle()) {
                this.deactivate(this.buttons.filter(current => current !== button));
                button.setSelected(true);
            }
            else {
                button.setSelected(!button.selected);
            }
            button.setAttribute(ariaChecked, button.selected.toString());
            button.setAttribute(tabindex, button.tabIndex.toString());
        });
    }
    ngAfterContentInit() {
        const isRadioGroup = this.isSelectionSingle();
        const buttonsRole = isRadioGroup ? 'radio' : 'checkbox';
        let anyChecked = false;
        this.buttons.forEach((button) => {
            button.setAttribute(ariaChecked, button.selected.toString());
            button.setAttribute(role, buttonsRole);
            if (isPresent(this.disabled)) {
                button.disabled = this.disabled;
            }
            if (this.look !== 'default') {
                button.look = this.look;
            }
            if (!isRadioGroup || button.selected) {
                button.setAttribute(tabindex, button.tabIndex.toString());
            }
            else if (isRadioGroup && !button.selected) {
                button.setAttribute(tabindex, "-1");
            }
            anyChecked = anyChecked || button.selected;
        });
        if (isRadioGroup && !anyChecked) {
            this.buttons.first.setAttribute(tabindex, this.buttons.first.tabIndex.toString());
            this.buttons.last.setAttribute(tabindex, this.buttons.last.tabIndex.toString());
        }
    }
    ngAfterViewChecked() {
        if (this.buttons.length) {
            this.buttons.first.renderer.addClass(this.buttons.first.element, 'k-group-start');
            this.buttons.last.renderer.addClass(this.buttons.last.element, 'k-group-end');
        }
    }
    ngOnDestroy() {
        this.subscription.unsubscribe();
    }
    ngAfterContentChecked() {
        this.verifySettings();
    }
    deactivate(buttons) {
        buttons.forEach((button) => {
            button.setSelected(false);
            button.setAttribute(ariaChecked, button.selected.toString());
            button.setAttribute(tabindex, "-1");
        });
    }
    activate(buttons) {
        buttons.forEach((button) => {
            button.setSelected(true);
            button.setAttribute(ariaChecked, button.selected.toString());
            button.setAttribute(tabindex, "0");
            button.focus();
        });
    }
    verifySettings() {
        if (isDevMode()) {
            if (this.isSelectionSingle() && this.buttons.filter(button => button.selected).length > 1) {
                throw new Error('Having multiple selected buttons with single selection mode is not supported');
            }
        }
    }
    isSelectionSingle() {
        return this.selection === 'single';
    }
};
tslib_1.__decorate([
    Input('disabled'),
    tslib_1.__metadata("design:type", Boolean)
], ButtonGroupComponent.prototype, "disabled", void 0);
tslib_1.__decorate([
    Input('selection'),
    tslib_1.__metadata("design:type", String)
], ButtonGroupComponent.prototype, "selection", void 0);
tslib_1.__decorate([
    Input('width'),
    tslib_1.__metadata("design:type", String)
], ButtonGroupComponent.prototype, "width", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", String)
], ButtonGroupComponent.prototype, "look", void 0);
tslib_1.__decorate([
    ContentChildren(ButtonDirective),
    tslib_1.__metadata("design:type", QueryList)
], ButtonGroupComponent.prototype, "buttons", void 0);
tslib_1.__decorate([
    HostBinding('class.k-button-group'),
    tslib_1.__metadata("design:type", Boolean),
    tslib_1.__metadata("design:paramtypes", [])
], ButtonGroupComponent.prototype, "wrapperClass", null);
tslib_1.__decorate([
    HostBinding('class.k-state-disabled'),
    tslib_1.__metadata("design:type", Boolean),
    tslib_1.__metadata("design:paramtypes", [])
], ButtonGroupComponent.prototype, "disabledClass", null);
tslib_1.__decorate([
    HostBinding('class.k-button-group-stretched'),
    tslib_1.__metadata("design:type", Boolean),
    tslib_1.__metadata("design:paramtypes", [])
], ButtonGroupComponent.prototype, "stretchedClass", null);
tslib_1.__decorate([
    HostBinding('class.k-button-group-bare'),
    tslib_1.__metadata("design:type", Boolean),
    tslib_1.__metadata("design:paramtypes", [])
], ButtonGroupComponent.prototype, "isBare", null);
tslib_1.__decorate([
    HostBinding('class.k-button-group-flat'),
    tslib_1.__metadata("design:type", Boolean),
    tslib_1.__metadata("design:paramtypes", [])
], ButtonGroupComponent.prototype, "isFlat", null);
tslib_1.__decorate([
    HostBinding('class.k-button-group-outline'),
    tslib_1.__metadata("design:type", Boolean),
    tslib_1.__metadata("design:paramtypes", [])
], ButtonGroupComponent.prototype, "isOutline", null);
tslib_1.__decorate([
    HostBinding('attr.role'),
    tslib_1.__metadata("design:type", String),
    tslib_1.__metadata("design:paramtypes", [])
], ButtonGroupComponent.prototype, "getRole", null);
tslib_1.__decorate([
    HostBinding('attr.dir'),
    tslib_1.__metadata("design:type", String),
    tslib_1.__metadata("design:paramtypes", [])
], ButtonGroupComponent.prototype, "dir", null);
tslib_1.__decorate([
    HostBinding('attr.aria-disalbed'),
    tslib_1.__metadata("design:type", Boolean),
    tslib_1.__metadata("design:paramtypes", [])
], ButtonGroupComponent.prototype, "ariaDisabled", null);
tslib_1.__decorate([
    HostBinding('style.width'),
    tslib_1.__metadata("design:type", String),
    tslib_1.__metadata("design:paramtypes", [])
], ButtonGroupComponent.prototype, "wrapperWidth", null);
tslib_1.__decorate([
    HostListener('keydown', ['$event']),
    tslib_1.__metadata("design:type", Function),
    tslib_1.__metadata("design:paramtypes", [Object]),
    tslib_1.__metadata("design:returntype", void 0)
], ButtonGroupComponent.prototype, "keydown", null);
ButtonGroupComponent = tslib_1.__decorate([
    Component({
        exportAs: 'kendoButtonGroup',
        providers: [KendoButtonService],
        selector: 'kendo-buttongroup',
        template: `
        <ng-content select="[kendoButton]"></ng-content>
    `
    }),
    tslib_1.__param(1, Optional()), tslib_1.__param(1, Inject(RTL)),
    tslib_1.__metadata("design:paramtypes", [KendoButtonService, Boolean])
], ButtonGroupComponent);
export { ButtonGroupComponent };
